python/ray/tune/execution/tune_controller.py

```diff
@@ -1605,6 +1605,29 @@ def _schedule_graceful_trial_stop(self, trial: Trial):
             self._schedule_trial_stop(trial)
 
     def _schedule_trial_pause(self, trial: Trial, should_checkpoint: bool = True):
+        if _use_storage_context():
+            if trial not in self._trial_to_actor:
+                logger.debug(
+                    f"Trial PAUSE requested for trial {trial} but trial is already "
+                    f"stopping. Ignoring."
+                )
+                return
+
+            if should_checkpoint:
+                # We need to wait for the save to finish before stopping the trial.
+                def stop_after_save_result(*args, **kwargs):
+                    self._on_saving_result(*args, **kwargs)
+                    self._schedule_trial_stop(trial)
+                    self._set_trial_status(trial, Trial.PAUSED)
```

Comment:
@krfricke Is it ok that there's some time between the schedule save and the save result coming in, where the trial is still `RUNNING`?

Previously, the memory checkpoint path would just return a future and immediately set to PAUSED.

Reply:
Ah good point.

I think it is a problem. If a pause instruction doesn't immediately pause the trial, we may send multiple instructions.

How about we set the status to PAUSED immediately, and again after the saving result resolved?

Reply:
Ok, done